### PR TITLE
Add Java GC Logs

### PR DESCRIPTION
**Pull request by MatthewKhouzam**
This feature is taken from the Eclipse JIFA project, the code was slightly tweaked, removing certain dependencies such as Lombok and SL4J, but it works well, is well tested and brings in several new views that are of interest:

* Java Memory VS Time, in the counters views, one can see the memory status by pool
* GC latency analysis: Shows the amount of time spent in the GC.

The code that will be kept if JIFA makes a jar for GC parsing would be GCTrace.java as it is the adapter between the GCModel and Trace Compass. Special attention in the reivew should be put there.

Known Issues:
* The GC log is entirely read to ram in the init phase. this can freeze the UI thread.
* The GC parser doesn't always read GC events as it changes from one JVM to another.

### CODE REVIEW

**Review by arfio — CHANGES_REQUESTED** [review]
Thank you for this addition that will improve java program performance analysis. Is it possible to add instructions on how to get a GC trace or give a link to those instructions ? Aside from that I put some small comments to improve maintainability.

File: tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/GCTrace.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        }
+        return 0;
+    }
+
+    @Override
+    public Iterable<ITmfEventAspect<?>> getEventAspects() {
+        if (fAspects.isEmpty()) {
+            List<ITmfEventAspect<?>> aspects = fAspects;
+            aspects.addAll((Collection<? extends ITmfEventAspect<?>>) super.getEventAspects());
+            aspects.add(2, simpleAspect("Allocation")); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (MatthewKhouzam) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/GCTrace.java:181]: It was for the order. Basically content is big and people ignore what comes after it.

[the remark this answers is not in the document]
Comment (MatthewKhouzam) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/GCTrace.java:181]: I am now using i++ all the way.

File: tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/GCTrace.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        return null;
+    }
+
+    private TmfEvent convert(GCEvent event) {
+
+        TmfEventType type = new TmfEventType(Objects.requireNonNull(event.getEventType().getName()), null);
+        GCModel model = Objects.requireNonNull(fModel);
+        ITmfTimestamp time = TmfTimestamp.fromNanos((long) ((event.getStartTime() + model.getReferenceTimestamp()) * 1e6));
+        List<TmfEventField> fields = new ArrayList<>();
+        fields.add(new TmfEventField("Allocation", event.getAllocation(), null)); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (MatthewKhouzam) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/GCTrace.java:141]: Done!

File: tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/xmlanalysis/gc_analysis.xml
[the remark this answers is not in the document]
Comment (MatthewKhouzam) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/xmlanalysis/gc_analysis.xml:1]: Yes, look in the tests plugin!

File: tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/util/Constant.java
[the remark this answers is not in the document]
Comment (MatthewKhouzam) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/util/Constant.java:1]: Done!

**Comment by MatthewKhouzam** [comment]
how to log in GC. I will add to wiki.

Java 8: Use -Xloggc:<file> -XX:+PrintGCDetails -XX:+PrintGCDateStamps.
Java 9+: Use -Xlog:gc*:file=<file> or more specific logging options.

File: tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/util/Constant.java
[the remark this answers is not in the document]
Comment (arfio) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/util/Constant.java:1]: is it ?

File: tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/xmlanalysis/gc_analysis.xml
[the remark this answers is not in the document]
Comment (arfio) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/xmlanalysis/gc_analysis.xml:1]: It seems that most of those traces are just samples and not a complete trace ?

File: tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/util/Constant.java
[the remark this answers is not in the document]
Comment (MatthewKhouzam) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/util/Constant.java:1]: this one is the code from Jifa. I ddin't modify it. Want it anyway?

[the remark this answers is not in the document]
Comment (arfio) [reply tracetypes/org.eclipse.tracecompass.incubator.jifa.core/src/org/eclipse/tracecompass/incubator/internal/jifa/core/gclog/util/Constant.java:1]: Yes, I think we should put the one they use for the other files.

**Review by arfio — APPROVED** [review]
LGTM !

**Comment by MatthewKhouzam** [comment]
Thanks!

**Pull request merged**